uhyve-interface/src/lib.rs

```diff
@@ -0,0 +1,112 @@
+//! # Uhyve Hypervisor Interface
+//!
+//! The uhyve hypercall interface works as follows:
+//!
+//! - On x86_64 you use a out port instruction. The address of the out_port corresponds to the
+//! hypercall you want to use. You can obtain it from the [`IoPorts`] enum. The data send to
+//! that port is the physical memory (in the VM) of the parameters of that hypercall.
+//! - On aarch64 you write to the respective [`HypercallAddress`]. The 64-bit data written to that location is the guest's physical memory address of the hypercall's parameter.
+
+#![no_std]
+
+// TODO: Throw this out, once https://github.com/rust-lang/rfcs/issues/2783 is resolved
+use num_enum::TryFromPrimitive;
+
+pub mod elf;
+pub mod parameters;
+use parameters::*;
+
+/// The version of the uhyve interface. Note: This is not the same as the semver of the crate but
+/// should be increased on every version bump.
+pub const UHYVE_INTERFACE_VERSION: u32 = 1;
+
+/// Enum containing all valid port mappings for hypercalls.
+///
+/// The discriminants of this enum are the respective ports, so one can get the code by calling
+/// e.g., `HypercallPorts::FileWrite as u16`.
+#[non_exhaustive]
+#[repr(u16)]
+#[derive(Debug, Eq, PartialEq, TryFromPrimitive)]
+pub enum HypercallAddress {
+	/// Port address = `0x400`
+	FileWrite = 0x400,
+	/// Port address = `0x440`
+	FileOpen = 0x440,
+	/// Port address = `0x480`
+	FileClose = 0x480,
+	/// Port address = `0x500`
+	FileRead = 0x500,
+	/// Port address = `0x540`
+	Exit = 0x540,
+	/// Port address = `0x580`
+	FileLseek = 0x580,
+	#[deprecated = "was never really in use"]
+	/// Port address = `0x640`
+	Netwrite = 0x640,
+	#[deprecated = "was never really in use"]
+	/// Port address = `0x680`
+	Netread = 0x680,
+	#[deprecated = "was never really in use"]
+	/// Port address = `0x700`
+	Netstat = 0x700,
+	/// Port address = `0x740`
+	Cmdsize = 0x740,
+	/// Port address = `0x780`
+	Cmdval = 0x780,
+	/// Port address = `0x800`
+	Uart = 0x800,
+	/// Port address = `0x840`
+	FileUnlink = 0x840,
+}
+impl From<Hypercall<'_>> for HypercallAddress {
+	fn from(value: Hypercall) -> Self {
+		match value {
+			Hypercall::Cmdsize(_) => Self::Cmdsize,
+			Hypercall::Cmdval(_) => Self::Cmdval,
+			Hypercall::Exit(_) => Self::Exit,
+			Hypercall::FileClose(_) => Self::FileClose,
+			Hypercall::FileLseek(_) => Self::FileLseek,
+			Hypercall::FileOpen(_) => Self::FileOpen,
+			Hypercall::FileRead(_) => Self::FileRead,
+			Hypercall::FileWrite(_) => Self::FileWrite,
+			Hypercall::FileUnlink(_) => Self::FileUnlink,
+			Hypercall::SerialWriteByte(_) => Self::Uart,
+		}
+	}
+}
```

Comment:
Is this actually used? If you plan to use it for the kernel (not sure if required), an `impl From<_> for &HypercallAddress` seems more useful because of moving the `&mut`.

Reply:
Note: `impl From<&Hypercall<'_>> for HypercallAddress`